### ISSUE TITLE
Certain tree updates can lead to call stack exhaustion with recursive `stack_node_add_link`

### ISSUE THREAD

**Issue by logancollins** [tree-sitter/tree-sitter#1939]
We're running into an issue with using the HTML, Ruby, and EmbeddedTemplate grammars together to mix HTML+ERB.

<img width="740" alt="Screen Shot 2022-10-31 at 12 41 20 PM" src="https://user-images.githubusercontent.com/714246/199096169-4b60f04e-3f11-4d96-9a94-86f4f2eb5939.png">

With certain tree updates to embedded Ruby in the ERB tags, the call stack eventually exhausts from too many recursive calls to `stack_node_add_link` in stack.c. I've been trying to trace exactly what parser rules or circumstances might be causing this, but it's been difficult to track down reliably.

Specifically, we're seeing it when users open a new ERB Ruby expression tag:

![2022-10-31 12 44 55](https://user-images.githubusercontent.com/714246/199096989-a38f3d1f-42f0-4d65-a9fa-39f3d51eaed5.gif)

I suppose my question: Do you think this is a reasonable case where the grammar is just causing _a huge number_ of recursive calls to this function and it's just exhausting the call stack, so tree-sitter shouldn't be using recursion; or is this more likely something going wrong in the grammar itself and there should never, ever be this many deep link updates in the first place?

I'm attaching the file we're working with, although the steps to set up environment are complex, since this is an injected language: [index.html.erb.zip](https://github.com/tree-sitter/tree-sitter/files/9904605/index.html.erb.zip).

This was reproduced in Nova 10, which uses tree-sitter's C library and native dylib compiled versions of the Tree Sitter core grammars (including HTML, Ruby, and EmbeddedTemplate).

If I can provide more specific details or help with diagnosis, let me know! I'm sort of out of my depth for figuring this out on my own.

Thank you for your assistance!

**Comment by maxbrunsfeld** [comment]
When the error happens, is it parsing the Ruby, the HTML, or the outer template?

**Comment by logancollins** [comment]
A couple more details in this case:

When this is parsing the Ruby injection layer, there's multiple (maybe 9?) included ranges being parsed, one for the contents of each ERB tag's ruby expression. Adding the new `<%` encompasses the entire rest of the file in one of these ranges that replaces everything else after it in the file. In case this has something to do with suddenly the contents of many disparate ranges in the parse tree being replaced with one.

**Comment by maxbrunsfeld** [comment]
If we can come up with a very clear reproduction case for this, I would definitely put some focused time toward fixing it very soon. Ideally, we'd come up with a way to cause the crash from either the `tree-sitter` CLI, or self-contained C/Rust program  (similar to the one [here](https://tree-sitter.github.io/tree-sitter/using-parsers#multi-language-documents)) that causes the crash.

I think it's very likely that there is some specific state of the buffer that will always cause this crash, regardless of what editing sequence got you into that state. It's also possible (though I think much less likely) that the crash is actually dependent on the preceding edit operation, so we'd need an initial state + an edit.

The fact that your current reproduction depends on ERB + language injections definitely makes it a bit more complicated to reproduce. One thought is that the `tree-sitter highlight` CLI command does perform language injections (though it doesn't handle edits), so you might try reproducing it with that first.

**Comment by logancollins** [comment]
Alright, I think I've built a somewhat simple case that demonstrates the issue as a C command line tool.

To get this set up:

1. Unpack this test project: [tree-sitter-crash-test.zip](https://github.com/tree-sitter/tree-sitter/files/9923343/tree-sitter-crash-test.zip)
2. In a terminal from the root unpacked directory, run `./build-tree-sitter.sh`.
  - This will download tree-sitter, the ruby grammar, and the embedded_template grammar and build all three as static libraries.
  - You may need to tweak the build flags in that shell script depending on your local development platform / environment (I'm on macOS).
3. Run `make` in the root directory.
  - This will build the test example, `crash-test`, and link against the built libraries from step 2.
4. Run `./crash-test`.
  - This should parse the file as ERB, enumerate matched code regions, then repairs as Ruby. That's when the infinite recursion should occur.

Some notes:
- I wasn't able to reproduce this in UTF-8 mode, only UTF-16, so it's possible there's something specific going on there.
- We have a custom parser input reader, so I tried to replicate that _very simply_ for a UTF-16 buffer rather than use the built-in one in tree-sitter, just in case.

**Comment by logancollins** [comment]
Not wanting to press the issue too much (as I'm sure you and others have a ton of stuff to work on), but I wondered if you had any idea on if a fix for this might be possible in the near-term. We're evaluating whether we want to revert our ERB syntax to using our regular expression parser instead of tree-sitter, as quite a few users are hitting this crash.

If the answer is "no," that's 100% okay! Just figuring out timelines for releases and stuff. It's always something we can come back to at a later date when a fix might be available.

Thanks again!

**Comment by maxbrunsfeld** [comment]
Ouch, really sorry to hear that many Nova users are getting crashes due to this.

I do have good news though - this morning I was just working on a fix related to this, because we are in the process of integrating ERB support into [Zed](https://zed.dev). Our randomized tests caught a bug that manifested with the same symptom: (a stack overflow in `stack_node_add_link` while editing ERB). I have fixed that bug, but haven't yet verified that my change fixes your reproduction case. I'll do that this afternoon.

**Comment by logancollins** [comment]
Oh, that's awesome to hear! No pressure at all, and thanks for helping out with this.

**Comment by maxbrunsfeld** [comment]
Actually, if I refer you to a certain commit, would you mind testing if it fixes your issue?

**Comment by logancollins** [comment]
Absolutely!

**Comment by maxbrunsfeld** [comment]
Ok, I have pushed the bugfix that I found to this branch: https://github.com/tree-sitter/tree-sitter/pull/1952. There are also a few other bugfixes on that branch related to the `ts_parser_set_included_ranges` API, which may be important for your use case in Nova.

That branch is close to being ready to merge - it's one fairly straightforward new API, and some bugfixes, but I am thinking about trying to add a bit more test coverage before merging it into master.
